### PR TITLE
Use rubex for faster regular expressions

### PR DESCRIPTION
**Pull request by vmarkovtsev**
I had to rewrite the non-greedy modifier because it is not supported by oniguruma

Alarma! It requires the oniguruma library installed, see https://github.com/moovweb/rubex

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
This does not pass the test suite. Some of the files are detected wrong.

This gives 2x speedup on the benchmark dataset and 1.6x speedup on the Linux kernel.

Máximo says that we could add an option in Makefile to build with oniguruma; it would rewrite the imports and several regexps which are incompatible.

Currently, I focus on passign the tests.

**Comment by vmarkovtsev** [comment]
I passed TestEnryTestSuite but am currently failing TestGeneratorTestSuite

**Comment by vmarkovtsev** [comment]
I passed the whole test suite. Now making the proper Travis config.

**Comment by mcuadros** [comment]
I found this interesting but always being a optional thing, and not the default, avoiding the requirement of cgo by default.

**Comment by vmarkovtsev** [comment]
@mcuadros This is not enabled by default, the import rewriting that you suggested is done with

```
make onigumura
```

and can be canceled with

```
make revert-onigumura
```

What's left in this PR is adding the documentation.

**Comment by vmarkovtsev** [comment]
All is done on my side, ready to the review.

File: README.md [hunk 1]
```diff
@@ -19,6 +19,31 @@ To build enry's CLI you must run
...
+sudo apt install libonig-dev
+```
+
+To build enry with Onigumura regexps, patch the imports with
+
+```
+make onigumura
+```
+
+nnd then rebuild the project.
```
Comment (bzz) [inline README.md]: and

File: README.md [hunk 2]
```diff
@@ -19,6 +19,31 @@ To build enry's CLI you must run
 it generates a binary in the project's root directory called `enry`. You can move this binary to anywhere in your `PATH`.
 
 
+### Faster regexp engine
+
+[Onigumura](https://github.com/kkos/oniguruma) is CRuby's regular expression engine.
```
Comment (bzz) [inline README.md]: Was confused reading this first time and had the same question as https://github.com/src-d/enry/pull/113#issuecomment-333099095

What do you think about highlighting here an optional nature of this improvement? 
Like`Optionally, you can enable [Origumara] - a CRuby's ..`

**Comment by vmarkovtsev** [comment]
Thanks @bzz fixed

File: Makefile [hunk 3]
```diff
@@ -58,6 +64,15 @@ benchmarks-slow: $(LINGUST_PATH)
 	mkdir -p benchmarks/output && go test -run=NONE -bench=. -slow -benchtime=100ms -timeout=100h >benchmarks/output/enry_samples.bench && \
 	benchmarks/linguist-samples.rb 5 >benchmarks/output/linguist_samples.bench
 
+$(RUBEX_ORIG): %.orig : %
+	sed -i.orig -e 's/"regexp"/regexp "github.com\/moovweb\/rubex"/g' $<
```
Comment (vmarkovtsev) [inline Makefile]: This is the universal sed syntax which works under macOS

File: .travis.yml [hunk 4]
```diff
@@ -4,16 +4,26 @@ go:
...
+  apt:
+    packages:
+      - libonig-dev
+
 matrix:
   allow_failures:
     - go: tip
   fast_finish: true
 
+env:
```
Comment (abeaumont) [inline .travis.yml]: Please rebase to master since with the publication of `enry-java`, it has changed quite a lot.

**Comment by vmarkovtsev** [comment]
@abeaumont Done, automatic rebase worked without conflicts, tests pass.

JDK jobs take ONIGUMURA=1, I don't think it's desired, right?

**Comment by abeaumont** [comment]
@vmarkovtsev right, ONIGUMURA should be 0 not only in JDK jobs but also in those jobs inside `jobs` section, that publish the shared libs, to be used by `enry-java`.

**Comment by vmarkovtsev** [comment]
fixed

**Pull request merged**